### ISSUE TITLE
Implement smarter equality checks for `GF`

### ISSUE THREAD

**Issue by Drake-Eidukas** [initc3/HoneyBadgerMPC#226]
Ran into this while debugging-- essentially, say you create two `GFElements` from two different `GF` objects, each of which are the exact same in every conceivable way except that the actual `GF` objects have a different internal `id` (simply different objects). You currently cannot perform math on these elements, as the operation fails when it checks for equality of the fields they were created on.

I propose we instead check equality of fields via checking their `modulus` field, as well as defining the `__hash__` method to explicitly be that of the `object` class.

@smkuls thoughts on this?

**Comment by Drake-Eidukas** [comment]
Essentially, in the `GFElement` class, when we try to do math operations, we check to see if the elements are from the same field via:
```
if self.field is not other.field:
    raise FieldsNotIdentical
```

This checks if the fields have the same id (i.e. same object), not whether or not the fields that the elements are from are the same (which would require comparing moduli rather than ids).

**Comment by Drake-Eidukas** [comment]
Alternatively, as pointed out by @smkuls, we should just implement `GF` using a more robust singleton pattern, for example, https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Singleton.html

**Comment by smkuls** [comment]
Nice catch, the intended functionality was supposed to be a `Singleton` and this implementation doesn't meet that criteria. The fix would be to make this a singleton class using the link pointed by @Drake-Eidukas and expose only the `get` method and prevent the constructor from getting called directly.

**Comment by Drake-Eidukas** [comment]
Closing issue as we merged in #229
